### ISSUE TITLE
Issues related to multiple field managers and group requests

### ISSUE THREAD

**Issue by bartgol** [E3SM-Project/scream#1093]
Currently, our AD is allowing any number of grids, and is storing a separate FM for each grid. The main time step loop always deals with reference grid fields, but some atm procs might want to save fields on non-ref grids, and possibly even restart from them.

The one usage of non-ref grid FM is currently given by dynamics. Homme stores the tracers at the end of the previous dyn step on the dyn grid, so it can use them to back out a tracers tendency. In other words, given an input Q on the ref grid, Homme does:

```
  Q_dyn = pd_remap(Q);
  FQ_dyn = (Q_dyn - Q_prev_dyn)/dt;
  Q_dyn = run_homme();
  Q_prev = Q_dyn;
  Q = dp_remap(Q_dyn)
```
An alternative approach would be to store Q_prev on the ref grid, and do
```
  FQ = (Q-Q_prev)/dt;
  FQ_dyn = pd_remap(FQ);
  Q_dyn = run_homme();
  Q = Q_prev = dp_remap(Q_dyn);
```
With a GLL physics grid, the two approaches would yield the same answer, but with pg2 they won't, since `pd(dp(X))!=X`.

Now, what's the issue? The issue comes when we have to register tracers_prev in the dyn grid FM. Homme wants bundled tracers, meaning that the groups `tracers` and `tracers_prev` need to be allocated as a single field. However, as of now, we have no way of ensuring that the order of the individual tracers in the group `tracers_prev` in the dyn grid FM is the same as the one in the group `tracers` in the ref grid FM. In other words, it could happen that we have `tracers = (qv, qr, qi)` on the ref grid, but `tracers_prev = (qv, qi, qr)` on the dyn grid.

Now, this is a very marginal problem, since dynamics doesn't really "know" what each tracer is, so the "label" we give to each slice is completely irrelevant (except qv, which has to be first, and we can currently do that with some hack). However, the feature of multiple FMs can give the false feeling that we can handle groups on multiple grids correctly, which might lead one day, say, radiation to store a copy of some of the tracers on the radiation-twin-columns grid (if we ever get to implement it), thinking that the remap phys->rad will handle it correclty, while the two might easily end up having tracers in different orders, which could cause tracer q2 to be remapped into the slot associated with tracer q3.

Now, my personal preference would be to prune non-ref grids FieldManager's from the AD completely. This would dramatically simplify some code in the AD cpp file, as well as in the FM class. Potentially, we could also remove the grid name from FieldIdentifier, FieldRequest, and GroupRequest classes, if deemed desirable. However, this would effectively prevent us from even saving some non-ref grid fields, just for output. So maybe a compromise solution could be to keep non-ref grid FMs in the AD, but to prohibit registration of field groups on non-ref grids (remember that the big hassle is just to link different groups).

The compromise solution would of course affect how dynamics handles Q_prev. In particular, dyn would have to use the second approach between the ones at the top of this post. As said above, this would not be bfb the same as what currently is done (which I believe is what eam does). OTOH, having dyn doing the backing out of FQ on the phys grid seems natural, since from the point of view of the AD _that's the only grid_ (all in-out fields of all atm procs are on the ref grid). And if one consider dyn as the black box `DYN = pd_remap + Homme + dp_remap`, then it makes sense to save and back out stuff on the ref grid.

I realize that doing this change would kind of remove some infrastructure that we spent time to develop. However, if the change would make things better, then so be it. Later on, we might realize we actually want this infrastructure back, so we might want to tag (in the git sense) the code, so that we can quickly revive the feature down the road.

@AaronDonahue, @PeterCaldwell, @tcclevenger , @jeff-cohere  ,I'd like to hear your thoughts, both in terms of algorithm, divergence from eam, and software design/maintainability.

P.s.: sorry for the very long post, and thanks for reading.

**Comment by jeff-cohere** [comment]
"Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away."
- Antoine de Saint-Exupery

I like the idea of simplifying and reducing the field managers and grids in the AD to accommodate the actual use cases we have to support. We've learned a lot by exploring more general solutions, but as @bartgol points out, we might be making promises that we can't keep. In particular, I think it would be easier for developers to focus their attention on a single grid/field manager within the AD. Those who deal with dynamics have to understand its pecularities in any case, so it doesn't bug me that dynamics is "special".

I don't know enough about how I/O works to speak to Luca's "compromise" proposal in which non-reference-grid field managers are kept within the AD but prohibited from using groups. Perhaps there's another solution that still allows us to jettison those things from the AD and still provide output. But even if not, the proposed compromise prevents developers from making assumptions about how groups work with these non-reference-grid fields, and that's a good thing.

If anyone can think of a specific situation in which we need more support for non-reference-grid field managers within the AD, now's the time to bring it up! Otherwise, I agree with Luca that we should tag our current implementation and strive for a single field manager in the AD.

**Comment by tcclevenger** [comment]
I like the compromise. Having `DYN = pd_remap + Homme + dp_remap` make things much easier to understand. How big of a deal is the non-bfbness?

**Comment by bartgol** [comment]
> I like the compromise. Having `DYN = pd_remap + Homme + dp_remap` make things much easier to understand. How big of a deal is the non-bfbness?

I don't really know if the fact that Homme saves Q_prev on dyn grid is _desired_ or it "simply happened to be implemented that way". It might also be that it has to do with something like reproducing a single time-step from homme standalone (which only has dyn grid).

Either way, I think that, conceptually speaking, it should be ok to store Q_prev on the phys grid (which we could just refer to as the "atm grid"), and see dyn as a pd+homme+dp black box.

**Comment by bartgol** [comment]
> "Perfection is achieved, not when there is nothing more to add, but when there is nothing left to take away."
> 
>     * Antoine de Saint-Exupery

Nice! Fits very well here.
 
> I don't know enough about how I/O works to speak to Luca's "compromise" proposal in which non-reference-grid field managers are kept within the AD but prohibited from using groups. Perhaps there's another solution that still allows us to jettison those things from the AD and still provide output. But even if not, the proposed compromise prevents developers from making assumptions about how groups work with these non-reference-grid fields, and that's a good thing.

The "compromise" solution would really be _just_ to allow dyn to save Q_prev on the dyn grid, like it's done today. However, I would argue for moving away from this seemingly odd choice, and save everything on one atm grid. If we do that, then we can really go for the "purge" solution.

**Comment by bartgol** [comment]
Update: from today's v1 telecon, I learned that we _must_ save/load Q_prev directly on dynamics grid, or we risk to lose resolution.

So I think we have two options:
- keep multiple grids
- have atm procs that need restart data on non-ref grid handle that internally.

The latter is not very satisfactory, since it puts a burdain on atm proc developers. In particular, they'd have to know how to use scorpio in/out classes, which conceptually they shouldn't have to.

So we're back to keeping multiple grids, and consequently multiple field managers. That's fine. What I would like to prioritize though is getting to work on #985 , which I will go and rephrase, since I think it's a bit old, and use language hinting at old data structure and/or design choices.

Once "internal" fields are available, we can add some peculiar logic to those, and those only. In particular, "internal" fields will be the only restart fields that can live on non-ref grids. All other fields needed for restart will be taken from the ref grid.

The main issue will remain: homme must ensure that Q_prev contains the same tracers in Q, and in the same order, even if nobody else registers fields on the dyn grid (let alone mark them as part of the "tracers" group). Which means that the infrastructure I added in the AD for group requests with "relative" group requests will still be needed, to some extent. However, I want to spend some time to see if it can be simplified, making the impl inside FieldManager a bit less involved.
